ACRIOS Systems

mgaspy2 opened the pull request:
This merge request introduces support for our M-Bus (Meter Bus) converter, which reads M-Bus payloads and sends them via LoRaWAN.

1. **Scheme**: We have prepared a scheme where the parsed M-Bus payload is stored as a single string property.
2. **Parser**: The custom parser reassembles fragmented M-Bus messages and parses the payload into a JSON format, which is then stored in the scheme's string property.

Please advise if the uplink.spec.js file is required for this integration, or if any further steps are necessary.

mariusgiger commented:
Hi @mgaspy2, thanks a lot for the contribution!

The changes look good and can soon be merged.

- Please rebase your branch to only include your changes. 
- Please add an `uplink.spec.js` with a test payload, you can find inspiration in other device types
- Please add an entry in `CHANGELOG.md` explaining your change

Cheers,
Marius

mgaspy2 commented:
Hello @mariusgiger,

Thank you for your feedback!

I have rebased our branch as requested, added the necessary changes, and included an entry in the CHANGELOG.md. 

I would like to clarify how to add multiple payloads and states to the uplink.spec.js file, as our parser is capable of assembling multiple fragmented payloads. Could you provide some guidance or examples on how best to structure those test cases?

Best regards,
Michal

mariusgiger commented:
Hi @mgaspy2, 

Thanks for the adaptations!

Is there a specific reason to output the result as a string (`parsed`)?  
Generally string notation is not the easiest way for data consumers to then use the data. Also akenza does not support visualizing a stringified object.

If possible output the data as an object

```json
emit("sample", {
  data: {
    len: 25,
    type: "Data",
    l: 19,
    c: 8,
    a: 5,
    ci: 115,
    errors: [],
    fixed: true,
    id: 12345678,
    accessN: 10,
    status: 0,
    cStored: "Actual",
    deviceCode: 7,
    deviceType: "Water meter",
    data: [
      {
        id: 0,
        storage: 0,
        func: "Instantaneous",
        value: 1,
        unit: "l",
      },
      {
        id: 1,
        storage: 1,
        func: "Instantaneous",
        value: 135,
        unit: "l",
      },
    ],
  },
  topic: "mbus",
});
```

Further if multiple measurements are included in the same uplink, you could emit multiple samples with different timestamps:

```
emit("sample", {
  data: {
    len: 25,
    type: "Data",
    l: 19,
    c: 8,
    a: 5,
    ci: 115,
    errors: [],
    fixed: true,
    id: 12345678,
    accessN: 10,
    status: 0,
    cStored: "Actual",
    deviceCode: 7,
    deviceType: "Water meter",
    id: 0,
    storage: 0,
    func: "Instantaneous",
    value: 1,
    unit: "l",
  },
  topic: "mbus",
  timestamp: new Date(),
});

emit("sample", {
  data: {
    len: 25,
    type: "Data",
    l: 19,
    c: 8,
    a: 5,
    ci: 115,
    errors: [],
    fixed: true,
    id: 12345678,
    accessN: 10,
    status: 0,
    cStored: "Actual",
    deviceCode: 7,
    deviceType: "Water meter",
    id: 1,
    storage: 1,
    func: "Instantaneous",
    value: 135,
    unit: "l",
  },
  topic: "mbus",
  timestamp: new Date(),
});

```

Currently, data can be easily directly visualized in akenza for flat objects and nested support will be added in the future.

Lastly, we recommend using speaking names rather than abbreviations (`l`, `c`, or `a`).

Cheers,
Marius

mgaspy2 commented:
Hi @mariusgiger,

Thanks for the clarification!

To address the concern, the reason we output the result as a string (parsed) is because every M-Bus device can have a unique payload structure, meaning the properties might appear in different orders or might be missing entirely depending on the device. Additionally, VIF/DIF filtering is applied on our device, which can modify the payload further by filtering out certain data elements before the final result is generated.

If I understand correctly, moving away from the string format would require us to implement specific schemas for every device our customers use, as well as for their customized filtered payloads. Given the variety and complexity of these devices, our solution is designed to support a wide range of M-Bus devices without needing to define rigid structures for each one.

Please let me know if you have any suggestions on how we might approach this differently while keeping the flexibility to support diverse M-Bus devices.

Best regards,
Michal

mariusgiger commented:
Hi @mgaspy2,

Thanks for the clarification - the schema is optional and helps to keep data consistent and visualize data.
You can omit it if you think it is tricky to create a generic schema to fit for all cases.

It is recommended to split data into separate topics if the structure changes in-between messages of the same device (uplinks). 

In any case, we recommend to output structured data (objects instead of strings) to make it easier to use by applications and plot it.

Cheers,
Marius

mgaspy2 commented:
Hi @mariusgiger ,

Thank you for the suggestion!

Given the variability in payloads from M-Bus devices, we need to keep the entire output under one topic to ensure consistency, as the structure can differ significantly. As you mentioned, creating a generic schema that fits all cases could be tricky, so we appreciate the flexibility.

The output from our parser is already in a dictionary JSON format. Is it possible to modify the schema to accommodate and display the output of our parser as it is, or would you recommend omitting the schema entirely in this case?

Looking forward to your advice!

Best regards,
Michal

mariusgiger commented:
Hi @mgaspy2,

I recommend to omit the schema entirely in this case and output the data as an object (not nested under parsed).

Best regards,
Marius

mgaspy2 commented:
Hello @mariusgiger,

Thank you for the recommendation. I’ve made the changes as suggested.

Best regards,
Michal

mariusgiger commented on types/acriosSystems/ACR-CV-101L-M-X/uplink.js:
```diff
@@ -0,0 +1,730 @@
+
+	if(thisFrameIndex == totalBatchFrames){
+		if(thisFrameIndex == 1) {
+			data = tmbus(thisFramePayload);
```
likely this needs to be changed to, as data needs to be an object

`data = JSON.parse(tmbus(thisFramePayload));`

mgaspy2 replied:
Hi @mariusgiger,

Thanks for pointing that out. The issue was that the parser was also emitting state and was still validating against the deleted schema. I forgot to modify this part when I removed the schema. I've made the necessary adjustments, and the test is now passing.

mariusgiger reviewed (commented):
Hi @mgaspy2, thanks for the adaptions - please make sure that the tests are working.

Cheers,
Marius
